ports/nrf/modules/machine/uart.c

```diff
@@ -182,8 +202,10 @@ STATIC mp_obj_t machine_hard_uart_make_new(const mp_obj_type_t *type, size_t n_a
     config.p_context = (void *)self;
 
     // Set NULL as callback function to keep it blocking
-    nrfx_uart_init(self->p_uart, &config, NULL);
-
+    nrfx_uart_init(self->p_uart, &config, m_rx_handler);
+    rx_ringbuf.iput = rx_ringbuf.iget = 0;
+    tx_pending = false;
+    nrfx_uart_rx(self->p_uart, &rx_buf[0], 1);
```

Comment:
I'm a bit unsure about this one, and it might be an edge case. The STARTRX task in the UART peripheral is issued in next line, this means that if there is a byte sent to UART as this point, it might not trigger an IRQ. So, if this is the trigger to get recursive RX IRQ's, maybe we should move this after RX has been enabled to be safe? I could check a bit up with the specification if the peripheral will trigger a IRQ on outstanding RX bytes.

Reply:
I'm not sure swapping the lines will make a difference because `rx_enable()` is anyway called by `nrfx_uart_rx()` if RX is not already enabled (see the check `if ((!p_cb->rx_enabled) && (!second_buffer))`).  So I guess we just remove the following `nrfx_uart_rx_enable()` altogether?

Reply:
At a first glance this looked correct to me, that we could remove the `nrfx_uart_rx_enable()` as the `rx_enable()` is done internally. 

However, looking over the code again, to me it looks like we have to call the `nrfx_uart_rx_enable()` once to get `p_cb->rx_enabled = true`. If not enabled (permanently false), `rx_enable()` will be called for all subsequent calls to `nrfx_uart_rx()`, which also indirectly clear the UART peripheral `NRF_UART_EVENT_RXDRDY` while interrupts for `NRF_UART_INT_MASK_RXDRDY` has been disabled (during entry of the `nrfx_uart_rx()` and activated again at exit).

Reply:
Ok, I see.  So should `nrfx_uart_rx_enable()` go before or after this line here?  I guess more it before?

Reply:
I'm not sure how many times i have dry-run this scenario in my head by now, but the conclusion is that the current code is the most correct. No re-order needed. 

An attempt on reasoning:

If done the other way around, `nrfx_uart_rx_enable()` before `nrf_uart_rx()`, we might get bytes into the UART HW buffers triggering an IRQ, but as there are no rx-buffer set at this point and it will result in an error instead and the bytes will be skipped.

So, next, why keep it. My current conclusion is that we would need `nrfx_uart_rx_enable()` to prevent `nrf_uart_rx()` to clear the `RXDRDY` issued from the callback handler, and prevent the UART IRQ handler to stop the RX flow (the nrfx UART driver IRQ handler expects to see the `RXDRDY` when the UART IRQ is hit in order to issue a callback to the application). In my head this could happen, but i might be wrong of course.

Reply:
Ok, I'll leave the order as-is.